allennlp_models/common/pretrained_model_config.py

```diff
@@ -0,0 +1,75 @@
+import os
+import weakref
+from dataclasses import dataclass
+from typing import Optional, Mapping, Dict
+from collections import defaultdict
+
+from allennlp.models import Model
+from allennlp.common.checks import ConfigurationError
+
+
+def get_description(model_class):
+    """
+    Returns the model's description from the docstring.
+    """
+    return model_class.__doc__.split("# Parameters")[0].strip()
+
+
+@dataclass(frozen=True)
+class ModelCard(object):
+    pass
+
+
+class PretrainedModelConfiguration(object):
+
+    __refs__ = defaultdict(list)  # type: ignore
+    _storage_location = "https://storage.googleapis.com/allennlp-public-models/"
+
+    def __init__(
+        self,
+        id: str,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        last_trained_on: Optional[str] = None,
+        archive_file: Optional[str] = None,
+        predictor_name: Optional[str] = None,
+        overrides: Optional[Mapping] = None,
+        model_card: Optional[ModelCard] = None,
+    ):
+
+        self.id = id
+        self.archive_file = archive_file
+        if self.archive_file and not self.archive_file.startswith("https:"):
+            self.archive_file = os.path.join(self._storage_location, self.archive_file)
+        self.name = name
+        self.description = description
+        self.last_trained_on = last_trained_on
+        self.predictor_name = predictor_name
+        self.overrides = overrides
+        self.model_card = model_card
+
+        self.__refs__[self.__class__].append(weakref.ref(self))
+
+    @staticmethod
+    def from_dict(config: Dict, model_class: Optional[Model] = None):
```

Comment:
There are two ways of making one of these, one with all the parameters (`__init__()`), and one here. Maybe you can combine them by adding a `model_class` parameter to `__init__()`, with a default value, so you never need to pass a dict that's basically just a list of parameters anyways? Does the dict serve another purpose that I haven't glimpsed yet?

Reply:
`model_class` should be `Optional[type]`. If it's `Optional[Model]`, that means it expects _instances_ of type `Model`.

I don't know if there is a way to restrict it more than `type`. It would be cool, but I don't know if that's possible in Python.